tests/vm_test.go

```diff
@@ -1921,8 +1924,67 @@ status:
 			}, 2*time.Minute, 1*time.Second).Should(Succeed())
 		})
 	})
+
+	Context("[Serial] when node becomes unhealthy", Serial, func() {
+		const componentName = "virt-handler"
+
+		It("[Serial] the VMs running in that node should be respawned", func() {
+			By("Starting VM")
+			vm := startVM(virtClient, createVM(virtClient, libvmi.NewCirros()))
+			vmi, err := virtClient.VirtualMachineInstance(vm.Namespace).Get(context.Background(), vm.Name, &k8smetav1.GetOptions{})
+			Expect(err).ToNot(HaveOccurred())
+
+			nodeName := vmi.Status.NodeName
+			oldUID := vmi.UID
+
+			By("Blocking virt-handler from reconciling the VMI")
+			libpod.AddKubernetesApiBlackhole(getHandlerNodePod(virtClient, nodeName), componentName)
+			Eventually(getHandlerNodePod(virtClient, nodeName).Items[0], 120*time.Second, time.Second, HaveConditionFalse(k8sv1.PodReady))
+
+			DeferCleanup(func() {
+				libpod.DeleteKubernetesApiBlackhole(getHandlerNodePod(virtClient, nodeName), componentName)
+				Eventually(getHandlerNodePod(virtClient, nodeName).Items[0], 120*time.Second, time.Second, HaveConditionTrue(k8sv1.PodReady))
+			})
+
+			pod, err := libvmi.GetPodByVirtualMachineInstance(vmi, vmi.Namespace)
+			Expect(err).ToNot(HaveOccurred())
+
+			By("Simulating loss of the virt-launcher")
+			err = virtClient.CoreV1().Pods(pod.Namespace).Delete(context.Background(), pod.Name, metav1.DeleteOptions{
+				GracePeriodSeconds: ptr.To(int64(0)),
+			})
+			Expect(err).ToNot(HaveOccurred())
+
+			// These timeouts are low on purpose. The VMI should be marked as Failed and be recreated fast.
+			// In case this fails, do not increase the timeouts.
+			By("The VM should not be Ready")
+			Eventually(ThisVM(vm), 30*time.Second, 1*time.Second).ShouldNot(beReady())
+
+			By("The controller should recreate the launcher pod and move the VMI to phase Scheduled")
+			Eventually(ThisVMI(vmi), 30*time.Second, 2*time.Second).Should(BeInPhase(v1.Scheduled))
```

Comment:
Can't this be flaky? or do we assume that the Pod will not move to `Running` within 2 seconds?

Reply:
If I'm not wrong, please correct otherwise, since the `Virt-handler` is not online at this point, the VMI can't transition to `Running`:

https://github.com/jcanocan/kubevirt/blob/a93aec57644a23189733f77a9e829a7fd2084e13/pkg/virt-handler/vm.go#L3307

It's true that it may happen that the VMI is rescheduled in another node and then moved to `Running` fast enough that the `Eventually` does not pick up this state. But that can be addressed with an affinity rule.

Reply:
We do not pin the VM into the specific node. Affinity seems like a hack, I think we can just `Eventually` check the `UID` as mentioned before, nothing more should be needed.

Reply:
Added `NodeSelector` rule.

Reply:
Ah okay. Sounds good to me.

Reply:
Dropped `Eventually(ThisVMI(vmi), 30*time.Second, 2*time.Second).Should(BeInPhase(v1.Scheduled))` and added `UID` `Eventually` check.